vignettes/measures-vignette.Rmd

```diff
@@ -0,0 +1,166 @@
+
+me = msrs(c("fairness.ppv", "fairness.npv"))
+predictions$score(me, task = adult_test) #Both should close to 1
+
+# Treatment equality
+# A classifier satisfies this definition if the subjects in the protected and unprotected groups have an equal ratio of FN and FP, satisfying the formula:
+# 
+# Assume A is the protected field with binary variable. Then FN{A=a}/FP{A=a} = FN{A=b}/FP{A=b}
+# 
+# However, we could do a simple transformation and assess FN{A=a}/FN{A=b} = FP{A=a}/FP{A=b}. Then we could use the following code:
+
+me = msrs(c("fairness.fp", "fairness.fn"))
+predictions$score(me, task = adult_test) #Measures should be close
+
+```
+
+# How to customize measures
+
+Users could see how to customize their measures from the following examples. Parameters users could define are base measures and operations.
+* To add a measure which only requires single metric, just create the MeasureFariness object
+* To add a measure which requires more metric, users are expected to add the base measures to mlr3_measures library first. Then create the fairness measure using `msrs()`.
+```{r}
+# False negative error rate balance, also referred to as equal opportunity.
+# A classifier satisfies equal opportunity if the subjects in the protected and unprotected groups have equal FNR. We could use the following code to assess it:
+
+me = MeasureFairness$new(operation = "groupwise_quotient", base_measure = msr("classif.fnr"))
+predictions$score(me, task = adult_test) #Should close to 1
+
+# Equalized Odds
+# Users are expected to evaluate Equalized Odds by ratio or absolute difference. A classifier satisfies Equalized Odds if the subjects in the protected and unprotected groups have equal TPR and equal FPR. So we could use the following code to assess Equalized Odds:
+
+x = getFromNamespace("mlr_measures", ns = "mlr3")
+x$add("fairness.fpr", MeasureFairness, base_measure = msr("classif.fpr"), operation = "groupwise_quotient")
+x$add("fairness.tpr", MeasureFairness, base_measure = msr("classif.tpr"), operation = "groupwise_quotient")
+predictions$score(msrs(c("fairness.fpr", "fairness.tpr")), task = adult_test) #Should close to 1
+```
+
+# How to compare the fairness performance between different learners using Benchmarks
+
+In this example, users could see how to create a benchmark object which compare the fairness measures between different learners and data tasks.
+
+```{r}
+design = benchmark_grid(
+  tasks = tsks(c("adult_train", "compas")),
+  learners = lrns(c("classif.ranger", "classif.rpart"),
+    predict_type = "prob", predict_sets = c("train", "test")),
+  resamplings = rsmps("cv", folds = 3)
+)
+
+bmr = benchmark(design)
+
+# Operations have been set to `groupwise_quotient()`
+measures = list( msr("fairness.tpr"), msr("fairness.npv"), msr("fairness.acc") )
```

Comment:
Also have a "classif.acc" for comparison

Reply:
Done